lib/ansible/module_utils/basic.py

```diff
@@ -2031,71 +2043,73 @@ def preexec():
-
-                if data:
-                    cmd.stdin.write(data)
-                    cmd.stdin.close()
-
-                while True:
-                    events = selector.select(1)
-                    for key, event in events:
-                        b_chunk = key.fileobj.read()
-                        if b_chunk == b(''):
-                            selector.unregister(key.fileobj)
-                        if key.fileobj == cmd.stdout:
-                            stdout += b_chunk
-                        elif key.fileobj == cmd.stderr:
-                            stderr += b_chunk
-                    # if we're checking for prompts, do it now
-                    if prompt_re:
-                        if prompt_re.search(stdout) and not data:
-                            if encoding:
-                                stdout = to_native(stdout, encoding=encoding, errors=errors)
-                            return (257, stdout, "A prompt was encountered while running a command, but no input data was specified")
-                    # only break out if no pipes are left to read or
-                    # the pipes are completely read and
-                    # the process is terminated
-                    if (not events or not selector.get_map()) and cmd.poll() is not None:
-                        break
-                    # No pipes are left to read but process is not yet terminated
-                    # Only then it is safe to wait for the process to be finished
-                    # NOTE: Actually cmd.poll() is always None here if no selectors are left
-                    elif not selector.get_map() and cmd.poll() is None:
-                        cmd.wait()
-                        # The process is terminated. Since no pipes to read from are
-                        # left, there is no need to call select() again.
-                        break
-
-                cmd.stdout.close()
-                cmd.stderr.close()
-                selector.close()
+            selector.register(cmd.stdout, selectors.EVENT_READ)
+            selector.register(cmd.stderr, selectors.EVENT_READ)
+
+            if os.name == 'posix':
+                fcntl.fcntl(cmd.stdout.fileno(), fcntl.F_SETFL, fcntl.fcntl(cmd.stdout.fileno(), fcntl.F_GETFL) | os.O_NONBLOCK)
+                fcntl.fcntl(cmd.stderr.fileno(), fcntl.F_SETFL, fcntl.fcntl(cmd.stderr.fileno(), fcntl.F_GETFL) | os.O_NONBLOCK)
+
+            if data:
+                cmd.stdin.write(data)
+                cmd.stdin.close()
+
+            while True:
+                # A timeout of 1 is both a little short and a little long.
+                # With None we could deadlock, with a lower value we would
+                # waste cycles. As it is, this is a mild inconvenience if
+                # we need to exit, and likely doesn't waste too many cycles
+                events = selector.select(1)
```

Comment:
Looks good to me to fix the hang, although I feel like the behavior on devel is more predictable, and that a script which wants to run something in the background should either redirect stdout and stderr, or be called with an async task. This/the previous behavior may include stdout/stderr from the background process, but only if it's noisy often enough.

Reply:
I wish either it would wait for none of the output from a background process (the default, probably what people expect) or all of it but I'm not sure how we'd accomplish the former even with a toggle.